cli/config.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package cli
+
+import (
+	_ "embed"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"github.com/sourcenetwork/defradb/logging"
+)
+
+// configPaths are config keys that will be made relative to the rootdir
+var configPaths = []string{
+	"datastore.badger.path",
+	"api.pubkeypath",
+	"api.privkeypath",
+}
+
+// configFlags is a mapping of config keys to cli flags to bind to.
+var configFlags = map[string]string{
+	"log.level":                         "loglevel",
+	"log.output":                        "logoutput",
+	"log.format":                        "logformat",
+	"log.stacktrace":                    "logtrace",
+	"log.nocolor":                       "lognocolor",
+	"api.address":                       "url",
+	"datastore.maxtxnretries":           "max-txn-retries",
+	"datastore.badger.inMemory":         "in-memory",
+	"datastore.badger.valuelogfilesize": "valuelogfilesize",
+	"net.peers":                         "peers",
+	"net.p2paddresses":                  "p2paddr",
+	"net.p2pdisabled":                   "no-p2p",
+	"api.allowed-origins":               "allowed-origins",
+	"api.pubkeypath":                    "pubkeypath",
+	"api.privkeypath":                   "privkeypath",
+}
+
+// defaultConfig returns a new config with default values.
+func defaultConfig() *viper.Viper {
+	cfg := viper.New()
+
+	cfg.AutomaticEnv()
+	cfg.SetEnvPrefix("DEFRA")
+	cfg.SetEnvKeyReplacer(strings.NewReplacer(".", "_", "-", "_"))
+
+	cfg.SetConfigName("config")
+	cfg.SetConfigType("yaml")
+
+	cfg.SetDefault("datastore.maxTxnRetries", 5)
+
+	cfg.SetDefault("datastore.badger.path", "data")
+	cfg.SetDefault("datastore.badger.inMemory", false)
+	cfg.SetDefault("datastore.badger.valueLogFileSize", 1<<30)
+
+	cfg.SetDefault("api.address", "127.0.0.1:9181")
+	cfg.SetDefault("api.pubKeyPath", "")
+	cfg.SetDefault("api.privKeyPath", "")
+	cfg.SetDefault("api.allowed-origins", []string{})
+
+	cfg.SetDefault("net.p2pDisabled", false)
+	cfg.SetDefault("net.p2pAddresses", []any{"/ip4/127.0.0.1/tcp/9171"})
+	cfg.SetDefault("net.pubSubEnabled", true)
+	cfg.SetDefault("net.relay", false)
+	cfg.SetDefault("net.peers", []string{})
+
+	cfg.SetDefault("log.level", "info")
+	cfg.SetDefault("log.stackTrace", true)
+	cfg.SetDefault("log.format", "csv")
+	cfg.SetDefault("log.output", "stderr")
+	cfg.SetDefault("log.noColor", false)
+	cfg.SetDefault("log.caller", false)
+
+	return cfg
+}
+
+// createConfig writes the default config file if one does not exist.
+func createConfig(rootdir string) error {
+	cfg := defaultConfig()
```

Comment:
suggestion: If you add the `bind cli flags to config keys` loop to `createConfig`, you will be able to remove all the `cfg.SetDefault` for the keys in `configFlags`. The flags default will be used instead and it will remove the need to maintain 2 sets of default values.

Reply:
I was hoping there was a way to do this but I didn't think of that. Thanks for the suggestion!